### PR TITLE
style(mypy): Enforcing typing for superset.db_engine_specs module

### PR DESCRIPTION
**Pull request by john-bodley**
### CATEGORY

Choose one

- [ ] Bug Fix
- [x] Enhancement (new features, refinement)
- [ ] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY

In an ideal world the entire the entire repo will be typed, though in order to achieve this small steps are required. This PR has type enforcement for the `superset.db_engine_specs` module (this was mostly typed and thus I felt this was a good place to start).

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS

to: @etr2460 @serenajiang @villebro @willbarrett

### CODE REVIEW

File: superset/db_engine_specs/base.py [hunk 1]
```diff
@@ -217,7 +225,7 @@ def make_select_compatible(
         return select_exprs
 
     @classmethod
-    def fetch_data(cls, cursor, limit: int) -> List[Tuple]:
+    def fetch_data(cls, cursor: Any, limit: int) -> List[Tuple]:
```
Comment (john-bodley) [inline superset/db_engine_specs/base.py]: The DBAPI cursor isn't wrapped by SQLALchemy and thus for now I used a generic type.

File: superset/db_engine_specs/base.py [hunk 2]
```diff
@@ -486,14 +496,13 @@ def extract_error_message(cls, e: Exception) -> str:
...
-    def _extract_error_message(cls, e: Exception) -> str:
+    def _extract_error_message(cls, e: Exception) -> Optional[str]:
         """Extract error message for queries"""
         return utils.error_msg_from_exception(e)
 
     @classmethod
-    def adjust_database_uri(cls, uri, selected_schema: Optional[str]):
-        """Based on a URI and selected schema, return a new URI
-
+    def adjust_database_uri(cls, uri: URL, selected_schema: Optional[str]) -> None:
```
Comment (john-bodley) [inline superset/db_engine_specs/base.py]: Given that we are mutating the `uri` object in-place, there's no need to return the object.

File: superset/db_engine_specs/presto.py [hunk 3]
```diff
@@ -670,10 +674,12 @@ def extra_table_metadata(
             col_names, latest_parts = cls.latest_partition(
                 table_name, schema_name, database, show_first=True
             )
-            latest_parts = latest_parts or tuple([None] * len(col_names))
+
+            if latest_parts:
+                latest_parts = [None] * len(col_names)  # type: ignore
```
Comment (john-bodley) [inline superset/db_engine_specs/presto.py]: I'm not sure if we're handling the latest parts in a consistent manner.

File: superset/utils/core.py [hunk 4]
```diff
@@ -430,10 +430,10 @@ def error_msg_from_exception(e):
     """
     msg = ""
     if hasattr(e, "message"):
-        if isinstance(e.message, dict):
-            msg = e.message.get("message")
-        elif e.message:
-            msg = e.message
+        if isinstance(e.message, dict):  # type: ignore
```
Comment (john-bodley) [inline superset/utils/core.py]: Workaround for https://github.com/python/mypy/issues/1424.

**Comment by willbarrett** [comment]
I am far from an expert on SQL types, but this looks pretty good! Thanks for adding types to this section of the codebase.

File: setup.cfg [hunk 5]
```diff
@@ -51,3 +51,9 @@ order_by_type = false
 
 [mypy]
 ignore_missing_imports = true
+
+[mypy-superset.db_engine_specs.*]
+check_untyped_defs = true
+disallow_untyped_calls = true
+disallow_untyped_defs = true
+no_implicit_optional = true
```
Comment (etr2460) [inline setup.cfg]: this is sick, i can't wait to add more of these going forward!

File: superset/db_engine_specs/base.py [hunk 6]
```diff
@@ -98,7 +104,7 @@ def _constructor(self):
 
 @compiles(TimestampExpression)
 def compile_timegrain_expression(
-    element: TimestampExpression, compiler: Compiled, **kw
+    element: TimestampExpression, compiler: Compiled, **kw: Any
```
Comment (etr2460) [inline superset/db_engine_specs/base.py]: while you're here, maybe we should change this to `kwargs` from `kw` to match the rest of the codebase

File: superset/db_engine_specs/base.py [hunk 7]
```diff
@@ -132,7 +138,9 @@ def get_allow_cost_estimate(cls, version: str = None) -> bool:
         return False
 
     @classmethod
-    def get_engine(cls, database, schema=None, source=None):
+    def get_engine(
+        cls, database: Database, schema: str = None, source: str = None
```
Comment (etr2460) [inline superset/db_engine_specs/base.py]: If you've enabled `no implicit optional` shouldn't these be `Optional[str]`?

File: superset/db_engine_specs/base.py [hunk 8]
```diff
@@ -77,7 +83,7 @@ class TimeGrain(NamedTuple):  # pylint: disable=too-few-public-methods
 class TimestampExpression(
     ColumnClause
 ):  # pylint: disable=abstract-method,too-many-ancestors,too-few-public-methods
-    def __init__(self, expr: str, col: ColumnClause, **kwargs):
+    def __init__(self, expr: str, col: ColumnClause, **kwargs: Any) -> None:
```
Comment (etr2460) [inline superset/db_engine_specs/base.py]: can't kwargs be a `Dict[String, Any]`? Maybe we should define a global `Kwargs` type that matches this

File: superset/db_engine_specs/base.py (hunk 2 above)
Comment (etr2460) [reply superset/db_engine_specs/base.py]: sometimes returning the object is nice so you can string together mutations on an object. Maybe we should keep returning it and type it appropriately?

**Review by etr2460 — COMMENTED** [review]
a few comments

File: superset/db_engine_specs/base.py (hunk 8 above)
Comment (serenajiang) [reply superset/db_engine_specs/base.py]: PEP484 says to type `**kwargs` and `*args` as the type of the arguments, not the type of `kwargs` (`dict`)  and `args`(`tuple`).

File: superset/db_engine_specs/base.py [hunk 9]
```diff
@@ -472,7 +482,7 @@ def get_all_datasource_names(
         return all_datasources
 
     @classmethod
-    def handle_cursor(cls, cursor, query, session):
+    def handle_cursor(cls, cursor: Any, query: Select, session: Session) -> None:
```
Comment (villebro) [inline superset/db_engine_specs/base.py]: I think here `query` is actually a `from superset.models.sql_lab import Query`

File: superset/db_engine_specs/base.py (hunk 2 above)
Comment (villebro) [reply superset/db_engine_specs/base.py]: If memory serves me right, these wouldn't benefit from chaining in the context they're used. However, I don't oppose returning the mutated object, but in that case I'd add a note in the docstring. Also there's a `# TODO:` below that can be removed.

File: superset/db_engine_specs/hive.py [hunk 10]
```diff
@@ -192,24 +195,23 @@ def convert_dttm(cls, target_type: str, dttm: datetime) -> Optional[str]:
         return None
 
     @classmethod
-    def adjust_database_uri(cls, uri, selected_schema=None):
+    def adjust_database_uri(cls, uri: URL, selected_schema: str = None) -> None:
```
Comment (villebro) [inline superset/db_engine_specs/hive.py]: I think `schema` type is missing `Optional[]`

File: superset/db_engine_specs/hive.py [hunk 11]
```diff
@@ -237,15 +239,17 @@ def progress(cls, log_lines):
...
         for line in log_lines:
             if lkp in line:
                 return line.split(lkp)[1]
-            return None
+        return None
 
     @classmethod
-    def handle_cursor(cls, cursor, query, session):  # pylint: disable=too-many-locals
+    def handle_cursor(  # pylint: disable=too-many-locals
+        cls, cursor: Any, query: Select, session: Session
```
Comment (villebro) [inline superset/db_engine_specs/hive.py]: Same as above

File: superset/db_engine_specs/presto.py [hunk 12]
```diff
@@ -712,7 +720,7 @@ def get_create_view(cls, database, schema: str, table: str) -> Optional[str]:
         return rows[0][0]
 
     @classmethod
-    def handle_cursor(cls, cursor, query, session):
+    def handle_cursor(cls, cursor: Any, query: Select, session: Session) -> None:
```
Comment (villebro) [inline superset/db_engine_specs/presto.py]: same as above

File: superset/db_engine_specs/snowflake.py [hunk 13]
```diff
@@ -47,14 +49,13 @@ class SnowflakeEngineSpec(PostgresBaseEngineSpec):
     }
 
     @classmethod
-    def adjust_database_uri(cls, uri, selected_schema=None):
+    def adjust_database_uri(cls, uri: URL, selected_schema: str = None) -> None:
```
Comment (villebro) [inline superset/db_engine_specs/snowflake.py]: Missing `Optional[]`

**Review by villebro — CHANGES_REQUESTED** [review]
First pass, small comments

File: superset/db_engine_specs/base.py (hunk 7 above)
Comment (john-bodley) [reply superset/db_engine_specs/base.py]: Agreed. I'm currently unsure why Mypy isn't enforcing that per [here](https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-no-implicit-optional).

**Comment by john-bodley** [comment]
Thanks @etr2460 and @villebro for the feedback. I've addressed your comments.

File: superset/db_engine_specs/snowflake.py (hunk 13 above)
Comment (villebro) [reply superset/db_engine_specs/snowflake.py]: This is still missing an `Optional`

**Review by villebro — CHANGES_REQUESTED** [review]
I think you missed one change, other than that LGTM

**Review by villebro — APPROVED** [review]

**Comment by john-bodley** [comment]
@villebro I updated the PR which leverages logic from https://github.com/apache/incubator-superset/pull/9150

**Comment by villebro** [comment]
@john-bodley noted; I already had the chance to re-review and approve. Btw, we're hoping to cut `0.36` soon, and to make sure future cherries apply as cleanly as possible, it would be good if this PR would make the cut. So best to merge this sooner than later.

**Pull request merged**